### PR TITLE
1208 update actor from integrationevent

### PR DESCRIPTION
**Pull request by x-platformcoder**
## Description

Updating marketparticipants data through the use of integrationevents from the marketparticipant domain, it will remove the necessity to use the manual synchronization of the marketparticipants data.

**Tech notes:**

It is assumed that we can continue with our current market participant data in charges and apply integration events on top of that.

Acceptance criteria:

- [x] Charges market participants are automatically updated when integration events are published by the Market participant domain.

- [x] Remove functionality for manually synchronizing data from Market participant database to Charges database.

- [x] Remove Postman request for synchronizing data

- [x] Update architecture diagram (charge and charge links) to reflect the Market Participant domain, our new function for consuming the integration events published.

### CODE REVIEW

File: build/infrastructure/main/kv-shared-resources.tf [hunk 1]
```diff
@@ -117,6 +117,16 @@ data "azurerm_key_vault_secret" "sbs_metering_point_created_sub_charges_name" {
   key_vault_id = data.azurerm_key_vault.kv_shared_resources.id
 }
 
+data "azurerm_key_vault_secret" "sbt_market_participant_changed_name" {
+  name         = "sbt-market-participant-changed-name"
+  key_vault_id = data.azurerm_key_vault.kv_shared_resources.id
+}
+
+data "azurerm_key_vault_secret" "sbs_market_participant_changed_to_charges_name" {
+  name         = "sbs-market-participant-changed-to-charges-name"
```
Comment (HenrikSommer) [inline build/infrastructure/main/kv-shared-resources.tf:126]: For other integration event subscriptions, we have used the naming convention "<from domain>-sub-<receiving domain>-name. Can we replace "to" with "sub"?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        {
+            _unitOfWork = unitOfWork;
+            _gridAreaRepository = gridAreaRepository;
+            _logger = loggerFactory.CreateLogger(nameof(GridAreaPersister));
+        }
+
+        public async Task PersistAsync(GridAreaChangedEvent gridAreaChangedEvent)
+        {
+            if (gridAreaChangedEvent is null)
+                throw new ArgumentNullException(nameof(gridAreaChangedEvent));
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs:43]:
```suggestion
            ArgumentNullException.ThrowIfNull(gridAreaChangedEvent);
            
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            var existingGridArea = await _gridAreaRepository.GetOrNullAsync(
+                gridAreaChangedEvent.Id).ConfigureAwait(false);
+            if (existingGridArea is null)
+            {
+                var gridArea = new GridArea(
+                    gridAreaChangedEvent.Id,
+                    gridAreaChangedEvent.GridAreaId);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs:50]:
```suggestion
                var gridArea = new GridArea(gridAreaChangedEvent.Id, gridAreaChangedEvent.GridAreaId);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                await _gridAreaRepository.AddAsync(gridArea).ConfigureAwait(false);
+                _logger.LogInformation("GridArea ID {GridAreaId} has been persisted", gridArea.Id);
+            }
+            else
+            {
+                existingGridArea.GridAccessProviderId = gridAreaChangedEvent.GridAreaId;
+                _logger.LogInformation(
+                    "GridArea ID '{GridAreaId}' has changed AreaAccessProviderId to '{GridAccessProviderId}'",
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs:58]:
```suggestion
                    "GridArea ID '{GridAreaId}' has changed GridAccessProviderId to '{GridAccessProviderId}'",
```

File: build/infrastructure/main/kv-shared-resources.tf (hunk 1 above)
Comment (HenrikSommer) [reply build/infrastructure/main/kv-shared-resources.tf:126]: Team Titans have chosen the naming convention for this subscription.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            _marketParticipantRepository = marketParticipantRepository;
+            _logger = loggerFactory.CreateLogger(nameof(MarketParticipantPersister));
+            _unitOfWork = unitOfWork;
+        }
+
+        public async Task PersistAsync(MarketParticipantChangedEvent marketParticipantChangedEvent)
+        {
+            if (marketParticipantChangedEvent is null)
+                throw new ArgumentNullException(nameof(marketParticipantChangedEvent));
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs:43]:
```suggestion
            ArgumentNullException.ThrowIfNull(marketParticipantChangedEvent);
            
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        private void UpdateMarketParticipant(
+            MarketParticipantChangedEvent marketParticipantChangedEvent,
+            MarketParticipant existingMarketParticipant,
+            MarketParticipantRole businessProcessRole)
+        {
+            existingMarketParticipant.IsActive = marketParticipantChangedEvent.IsActive;
+            _logger.LogInformation(
+                "Marketparticipant ID '{MarketParticipantId}' with role '{BusinessProcessRole}' has changed state",
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs:75]:
```suggestion
                "Market participant with ID '{MarketParticipantId}' and role '{BusinessProcessRole}' has changed state",
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            var marketParticipant = new MarketParticipant(
+                Guid.NewGuid(),
+                marketParticipantChangedEvent.MarketParticipantId,
+                marketParticipantChangedEvent.IsActive,
+                businessProcessRole);
+
+            await _marketParticipantRepository.AddAsync(marketParticipant).ConfigureAwait(false);
+            _logger.LogInformation(
+                "Marketparticipant ID '{MarketParticipantId}' with role '{BusinessProcessRole}' has been persisted",
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs:92]:
```suggestion
                "Market participant with ID '{MarketParticipantId}' and role '{BusinessProcessRole}' has been persisted",
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs (hunk 3 above)
Comment (x-platformcoder) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs (hunk 4 above)
Comment (x-platformcoder) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GridAreas/Handlers/GridAreaPersister.cs]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs (hunk 5 above)
Comment (x-platformcoder) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/EndpointTests/GridAreaPersisterEndPointTests.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            }
+
+            public Task InitializeAsync()
+            {
+                return Task.CompletedTask;
+            }
+
+            public Task DisposeAsync()
+            {
+                Fixture.MessageHubMock.Clear();
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/EndpointTests/GridAreaPersisterEndPointTests.cs:59]: MessageHub is not in use in the test-method.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/EndpointTests/MarketParticipantPersisterEndpointTests.cs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            }
+
+            public Task InitializeAsync()
+            {
+                return Task.CompletedTask;
+            }
+
+            public Task DisposeAsync()
+            {
+                Fixture.MessageHubMock.Clear();
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/EndpointTests/MarketParticipantPersisterEndpointTests.cs:53]: MessageHub is not in use in the test method.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/EndpointTests/GridAreaPersisterEndPointTests.cs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+                context.MarketParticipants.Add(markedParticipant);
+                await context.SaveChangesAsync().ConfigureAwait(false);
+                return markedParticipant.Id;
+            }
+
+            private static ServiceBusMessage CreateServiceBusMessage(
+                Guid id,
+                Guid gridAreaId,
+                out string correlationId,
+                out string parentId)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/EndpointTests/GridAreaPersisterEndPointTests.cs:102]: Personally, I prefer `tuples` over `out var` in a case like this and a "result"-class over `tubles` when more objects need be returned.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/EndpointTests/MarketParticipantPersisterEndpointTests.cs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                    out var correlationId,
+                    out var parentId);
+
+                // Act
+                await MockTelemetryClient.WrappedOperationWithTelemetryDependencyInformationAsync(
+                    () => Fixture.MarketParticipantChangedTopic.SenderClient.SendMessageAsync(message), correlationId, parentId);
+
+                // Assert
+                await FunctionAsserts.AssertHasExecutedAsync(Fixture.HostManager, nameof(MarketParticipantEndpoint)).ConfigureAwait(false);
+                await using var context = Fixture.DatabaseManager.CreateDbContext();
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/EndpointTests/MarketParticipantPersisterEndpointTests.cs:78]: Move to `// Arrange`?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantEndpoint.cs [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+                 case GridAreaUpdatedIntegrationEvent gridAreaEvent:
+                     {
+                         var gridAreaChangedEvent =
+                             MarketParticipantChangedEventMapper.MapFromGridArea(gridAreaEvent);
+                         await _gridAreaPersister
+                             .PersistAsync(gridAreaChangedEvent)
+                             .ConfigureAwait(false);
+                         break;
+                     }
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantEndpoint.cs:74]: Let's move this part to a handler class, that we can test automatically, like we do for example with `CreateChargeLinkReceiverEndpoint` and `CreateLinkRequestHandler`.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/IChargeFactory.cs [hunk 13]
```diff
@@ -14,6 +14,7 @@
 
 using System.Threading.Tasks;
 using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands;
+using GreenEnergyHub.Charges.Domain.MarketParticipants;
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/IChargeFactory.cs:17]: can be removed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantEndpoint.cs [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+            _marketParticipantEventHandler = marketParticipantEventHandler;
+        }
+
+        [Function(FunctionName)]
+        public async Task RunAsync(
+            [ServiceBusTrigger(
+                "%" + EnvironmentSettingNames.MarketParticipantChangedTopicName + "%",
+                "%" + EnvironmentSettingNames.MarketParticipantChangedSubscriptionName + "%",
+                Connection = EnvironmentSettingNames.DataHubListenerConnectionString)]
+            [NotNull] byte[] message)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantEndpoint.cs:44]: The [NotNull] can be removed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantEndpoint.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+using GreenEnergyHub.Charges.Application.MarketParticipants.Handlers;
+using GreenEnergyHub.Charges.FunctionHost.Common;
+using JetBrains.Annotations;
+using Microsoft.Azure.Functions.Worker;
+
+namespace GreenEnergyHub.Charges.FunctionHost.MarketParticipant
+{
+    public class MarketParticipantEndpoint
+    {
+        public const string FunctionName = nameof(MarketParticipantEndpoint);
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantEndpoint.cs:26]:
```suggestion
        private const string FunctionName = nameof(MarketParticipantEndpoint);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantChangedEventMapper.cs [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+// limitations under the License.
+
+using System.Linq;
+using Energinet.DataHub.MarketParticipant.Integration.Model.Dtos;
+using GreenEnergyHub.Charges.Domain.Dtos.GridAreas;
+using GreenEnergyHub.Charges.Domain.Dtos.MarketParticipantsChangedEvents;
+
+namespace GreenEnergyHub.Charges.Application.MarketParticipants.Handlers
+{
+    public static class MarketParticipantChangedEventMapper
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantChangedEventMapper.cs:22]: As this handles different types of events, I suggest to rename like this
```suggestion
    public static class MarketParticipantDomainEventMapper
```

**Review by prtandrup — COMMENTED** [review]
Minor comments.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantDomainEventMapper.cs [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+using GreenEnergyHub.Charges.Domain.Dtos.MarketParticipantsUpdatedEvents;
+
+namespace GreenEnergyHub.Charges.Application.MarketParticipants.Handlers
+{
+    public static class MarketParticipantDomainEventMapper
+    {
+        public static MarketParticipantUpdatedEvent MapFromActor(
+            ActorUpdatedIntegrationEvent actorUpdatedIntegrationEvent)
+        {
+            var isActive = actorUpdatedIntegrationEvent.Status is ActorStatus.Active or ActorStatus.New;
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantDomainEventMapper.cs:27]: From Charges' perspective `ActorStatus.New` is not to be considered isActive. New is considered a draft, and we should not send charge notifications to organizations in status new.

I'll remove it, but just wanted to note it here :)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantDomainEventMapper.cs [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+// limitations under the License.
+
+using System.Linq;
+using Energinet.DataHub.MarketParticipant.Integration.Model.Dtos;
+using GreenEnergyHub.Charges.Domain.Dtos.GridAreas;
+using GreenEnergyHub.Charges.Domain.Dtos.MarketParticipantsUpdatedEvents;
+
+namespace GreenEnergyHub.Charges.Application.MarketParticipants.Handlers
+{
+    public static class MarketParticipantDomainEventMapper
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantDomainEventMapper.cs:22]: We should add tests for this

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+                existingMarketParticipant.MarketParticipantId,
+                existingMarketParticipant.BusinessProcessRole);
+        }
+
+        private async Task AddMarketParticipantAsync(
+            MarketParticipantUpdatedEvent marketParticipantUpdatedEvent,
+            MarketParticipantRole businessProcessRole)
+        {
+            var marketParticipant = new MarketParticipant(
+                Guid.NewGuid(),
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs:87]: Something tells me that this is incorrect. I believe we need the ActorId of the `ActorUpdatedIntegrationEvent` instead of generating our own guid.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/MarketParticipantRepository.cs [hunk 20]
```diff
@@ -31,43 +31,79 @@ public class MarketParticipantRepository : IMarketParticipantRepository
...
+            MarketParticipantRole businessProcessRole,
+            string marketParticipantId)
         {
             return await _chargesDatabaseContext
                 .MarketParticipants
-                .SingleAsync(mp => mp.Id == id)
+                .SingleOrDefaultAsync(mp =>
+                    mp.MarketParticipantId == marketParticipantId && mp.IsActive &&
+                    mp.BusinessProcessRole == businessProcessRole)
                 .ConfigureAwait(false);
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/MarketParticipantRepository.cs:107]: I think `mp.IsActive` should be removed here, otherwise we won't be able to update the IsActive state?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/GridAreaLinkRepository.cs [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+            return await Task.FromResult(_chargesDatabaseContext
+                .GridAreaLinks
+                .SingleOrDefault(gal => gal.Id == gridAreaLinkId)).ConfigureAwait(false);
+        }
+
+        public async Task<GridAreaLink?> GetGridAreaOrNullAsync(Guid gridAreaId)
+        {
+            return await Task.FromResult(_chargesDatabaseContext
+                .GridAreaLinks
+                .Single(gal => gal.GridAreaId == gridAreaId)).ConfigureAwait(false);
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/GridAreaLinkRepository.cs:48]: above method uses .SingleOrDefault can we align?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+            MarketParticipantUpdatedEvent marketParticipantUpdatedEvent,
+            MarketParticipant existingMarketParticipant)
+        {
+            existingMarketParticipant.IsActive = marketParticipantUpdatedEvent.IsActive;
+            _logger.LogInformation(
+                "Market participant with ID '{MarketParticipantId}' and role '{BusinessProcessRole}' " +
+                "has changed state",
+                existingMarketParticipant.MarketParticipantId,
+                existingMarketParticipant.BusinessProcessRole);
+        }
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantPersister.cs:80]: I think we should add `await ConnectToGridAreaAsync` here as well. Otherwise we will miss the addition of new grid areas to an existing market participant?

**Review by prtandrup — COMMENTED** [review]
A couple of comments

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/GridAreaLinks/IGridAreaLinkRepository.cs [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+        /// Retrieves a grid area link
+        /// </summary>
+        /// <param name="gridAreaLinkId"></param>
+        Task<GridAreaLink?> GetOrNullAsync(Guid gridAreaLinkId);
+
+        /// <summary>
+        /// Retrieves a grid area link from grid area
+        /// </summary>
+        /// <param name="gridAreaId"></param>
+        Task<GridAreaLink?> GetGridAreaOrNullAsync(Guid gridAreaId);
```
[the remark this answers is not in the document]
Comment (x-platformcoder) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/GridAreaLinks/IGridAreaLinkRepository.cs:41]: the method signature is identical, so we need to have different names for the methods... or do I misunderstand your suggestion?

[the remark this answers is not in the document]
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/GridAreaLinks/IGridAreaLinkRepository.cs:41]: Nevermind :)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MarketParticipants/IMarketParticipantRepository.cs [hunk 24]
```diff
@@ -23,23 +23,42 @@ namespace GreenEnergyHub.Charges.Domain.MarketParticipants
     /// </summary>
     public interface IMarketParticipantRepository
     {
+        /// <summary>
+        /// Adds a new market participant
+        /// </summary>
+        /// <param name="marketParticipant"></param>
+        Task AddAsync(MarketParticipant marketParticipant);
+
         Task<MarketParticipant> SingleAsync(string marketParticipantId);
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MarketParticipants/IMarketParticipantRepository.cs:32]: Should we remove this method, and replace its' usages and implementation with `Task<MarketParticipant?> SingleOrNullAsync(string marketParticipantId);` to cope with the possible null?

Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MarketParticipants/IMarketParticipantRepository.cs:32]: Should we remove this method, and replace its' usages and implementation with `Task<MarketParticipant?> SingleOrNullAsync(string marketParticipantId);` to cope with the possible null?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationEventHandler.cs [hunk 25]
```diff
@@ -177,12 +177,12 @@ private static OperationType GetOperationType(ChargeOperationDto chargeOperation
         private async Task<Charge?> GetChargeAsync(ChargeOperationDto chargeOperationDto)
         {
             var marketParticipant = await _marketParticipantRepository
-                .SingleAsync(chargeOperationDto.ChargeOwner)
+                .SingleOrNullAsync(chargeOperationDto.ChargeOwner)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeInformationEventHandler.cs:180]: Let's roll back and rely on [1418](https://app.zenhub.com/workspaces/team-volt-60a61059028fd7000e49956e/issues/energinet-datahub/geh-charges/1418)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs [hunk 26]
```diff
@@ -112,12 +112,12 @@ public async Task HandleAsync(ChargeCommandReceivedEvent commandReceivedEvent)
         private async Task<Charge?> GetChargeAsync(ChargeOperationDto operation)
         {
             var marketParticipant = await _marketParticipantRepository
-                .SingleAsync(operation.ChargeOwner)
+                .SingleOrNullAsync(operation.ChargeOwner)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs:115]: Let's rollback and rely on [1419](https://app.zenhub.com/workspaces/team-volt-60a61059028fd7000e49956e/issues/energinet-datahub/geh-charges/1419)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/GridAreaLinkPersister.cs [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+                _logger.LogInformation(
+                    "GridAreaLink ID {GridAreaLink} for GridArea ID {GridAreaId} has been persisted",
+                    gridAreaLink.Id,
+                    gridAreaLink.GridAreaId);
+            }
+            else
+            {
+                if (existingGridAreaLink.GridAreaId == gridAreaUpdatedEvent.GridAreaId)
+                {
+                    return;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/GridAreaLinkPersister.cs:65]: Should we add a test for the event of equal existing and updated Id?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MarketParticipants/IMarketParticipantRepository.cs [hunk 28]
```diff
@@ -23,23 +24,42 @@ namespace GreenEnergyHub.Charges.Domain.MarketParticipants
...
+        /// Get the grid access provider from primary key
+        /// </summary>
+        Task<MarketParticipant?> GetGridAccessProviderAsync(Guid gridAreaId);
 
         /// <summary>
         /// Get the grid access provider of the grid area that the metering point belongs to.
         /// </summary>
         Task<MarketParticipant> GetGridAccessProviderAsync(string meteringPointId);
 
-        Task<List<MarketParticipant>> GetGridAccessProvidersAsync();
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MarketParticipants/IMarketParticipantRepository.cs:41]: Why was this removed? Its implemented in `MarketParticipantRepository`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/MarketParticipantRepository.cs [hunk 29]
```diff
@@ -81,6 +117,15 @@ public async Task<IReadOnlyCollection<MarketParticipant>> GetAsync(IEnumerable<G
...
+            return _chargesDatabaseContext
+                .MarketParticipants
+                .Where(mp => mp.BusinessProcessRole == MarketParticipantRole.GridAccessProvider)
+                .Where(m => m.IsActive)
+                .ToListAsync();
+        }
+
         public Task<MarketParticipant> GetGridAccessProviderAsync(string meteringPointId)
         {
             if (meteringPointId == null) throw new ArgumentNullException(nameof(meteringPointId));
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/MarketParticipantRepository.cs:131]:
```suggestion
            ArgumentNullException.ThrowIfNull(meteringPointId);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MarketParticipants/IMarketParticipantRepository.cs (hunk 28 above)
Comment (x-platformcoder) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MarketParticipants/IMarketParticipantRepository.cs:41]: It is still there but now in line 51

**Review by HenrikSommer — APPROVED** [review]

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/MarketParticipants/Handlers/MarketParticipantDomainEventMapperTests.cs [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+            var gridAreaUpdatedIntegrationEvent = new GridAreaUpdatedIntegrationEvent(
+                System.Guid.NewGuid(),
+                gridAreaId,
+                string.Empty,
+                string.Empty,
+                PriceAreaCode.DK1,
+                gridAreaLinkId);
+
+            // Act
+            var actualMarketParticipantUpdatedEvent =
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/MarketParticipants/Handlers/MarketParticipantDomainEventMapperTests.cs:80]:
```suggestion
            var actualGridAreaUpdatedEvent =
```

**Review by prtandrup — COMMENTED** [review]
minor things
